### PR TITLE
Documentation of the code for Doxygen

### PR DESCRIPTION
**Pull request by piyueh**
I finished the documentation for the purpose of creating HTML manuals (API manual and user manual) with Doxygen. To create and see the resulting HTML pages, go to the `doc` folder and run
```shell
$ doxygen Doxyfile
```
and use any preferred browser to open `doc/html/index.html`.

The documentation for API users are written in header files, and they are done with Doxygen command. I haven't put much effort to document the implementations in source files (i.e., *.cpp) because the documentation in the implementations is supposed to be for PetIBM developers instead of PetIBM users and is not a priority now.

All stuff in PetIBM is categorized into two top modules in the HTML manual. Building components (stuff in `src` folder) belong to on module, while applications (stuff in `applications` folder) belong to the other one.

Click the `Modules` link in the sidebar of the HTML manuals to see the two modules and their sub-modules.

### CODE REVIEW

**Comment by mesnardo** [comment]
@piyueh, I opened a PR to your PetIBM fork to the branch `factorization` to correct typos in the docstrings.
If it is ok with you, you can merge it and the commit will appear in the present PR.

**Comment by piyueh** [comment]
I've merged the PR to `piyueh/factorization`. Thanks!

**Comment by piyueh** [comment]
@mesnardo Seems this PR can be merged automatically without conflictions.

**Comment by mesnardo** [comment]
Ok, I'll finish the review very soon.

File: doc/DoxygenLayout.xml [hunk 1]
```diff
@@ -0,0 +1,204 @@
+<doxygenlayout version="1.0">
+  <!-- Generated by doxygen 1.8.13 -->
+  <!-- Navigation index tabs for HTML output -->
+  <navindex>
+    <tab type="mainpage" visible="yes" title=""/>
+    <tab type="user" url="@ref md_CHANGELOG" title="Change Logs"/>
```
Comment (mesnardo) [inline doc/DoxygenLayout.xml]: "Change Logs" -> "Change Log"

**Comment by mesnardo** [comment]
@piyueh, can you add a title to the markdown files that are in `doc/markdowns`?
For example: add `# Input files` to `inputs.md`.
This way, Doxygen will display the title and not the name of the file.

File: include/petibm/type.h [hunk 2]
```diff
@@ -18,232 +19,155 @@
 # include <petscmat.h>
 
 
-/** \brief operator < for MatStencil for using it as a key to map. 
+/** 
+ * \defgroup petibm Building blocks for building flow solvers
```
Comment (mesnardo) [inline include/petibm/type.h]: I propose to use "PetIBM building blocks" to avoid the repetition.

**Comment by piyueh** [comment]
@mesnardo I've made changes according to the review.

**Pull request merged**